### PR TITLE
feat(release): publish pipeline for wxcc widgets

### PR DESCRIPTION
**Pull request by Shreyas281299**
# This PR completes https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-575855

## What all changed?
- Added a new publish pipeline for cc-widgets
- Updated circleci to run install, linter, build and unit test for PRs
- Added tooling for publish

## Testing
- Check the commit history for my test branch - https://github.com/webex/widgets/tree/eft 
   - Made a patch release
   - Made a minor release
## Notes
- For tooling the tests ive written are 100% coverage except that 1 line, its not possible to test with jest. But apart from that line everything else should be covered.
Screenshot 2024-11-20 at 1 37 25 PM

- New deploy pipeline for cc-widgets
Screenshot 2024-11-20 at 2 00 23 PM

- New PR pipeline for PRs
Screenshot 2024-11-20 at 2 36 16 PM

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -37,7 +37,7 @@ jobs:
           root: .
           paths:
             - node_modules
-            - packages/*/node_modules
```
Comment (Shreyas281299) [inline .circleci/config.yml:40]: We have changed the file structure packages will be inside package/**/*
Examples : 
packages/contact-center/node_modules
packages/widgets/node_modules (https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-583602)

File: .circleci/config.yml [hunk 2]
```diff
@@ -87,6 +90,25 @@ jobs:
           name: Release
           command: yarn workspace @webex/widgets run release
 
+  unit_tests:
```
Comment (Shreyas281299) [inline .circleci/config.yml:93]: Added unit test step in pipeline. We test tooling first and then cc-widgets.

File: .circleci/config.yml [hunk 3]
```diff
@@ -87,6 +90,25 @@ jobs:
...
+    steps:
+      - checkout_from_workspace
+      - run:
+          name: Test Tooling
+          command: yarn run test:tooling
+      - run:
+          name: Test CC Widgets
+          command: yarn run test:cc-widgets
+  
+  release_widgets:
```
Comment (Shreyas281299) [inline .circleci/config.yml:104]: This step will only release cc-widgets

File: .circleci/config.yml [hunk 4]
```diff
@@ -98,19 +120,32 @@ workflows:
       - build:
           requires:
             - linter
-          filters:
-            branches:
-              only: master
+
```
Comment (Shreyas281299) [inline .circleci/config.yml:123]: By removing the filter we make sure that build step runs on PRs as well.
Screenshot 2024-11-20 at 2 36 16 PM

File: package.json [hunk 5]
```diff
@@ -11,20 +11,38 @@
...
     "samples:serve": "open docs/index.html && yarn workspace @webex/react-samples-app serve",
     "samples:serve-react": "yarn workspace @webex/react-samples-app serve",
-    "samples:serve-wc": "yarn workspace @webex/web-component-samples-app serve"
+    "samples:serve-wc": "yarn workspace @webex/web-component-samples-app serve",
+    "release:widgets": "semantic-release"
+  },
+  "jest": {
+    "testMatch": [
+      "**/tooling/tests/*.js"
+    ]
```
Comment (Shreyas281299) [inline package.json:46]: When we run `yarn run test:tooling' this is what make sure we only run tooling.

File: packages/contact-center/station-login/package.json [hunk 6]
```diff
@@ -3,6 +3,13 @@
   "description": "Webex Contact Center Widgets: Station Login",
   "version": "1.0.0",
   "main": "dist/index.js",
+  "publishConfig": {
+    "access": "public"
+  },
+  "files": [
+    "dist/",
+    "package.json"
+  ],
```
Comment (Shreyas281299) [inline packages/contact-center/station-login/package.json:12]: This ensures that we only publish the dist and package.json to npmjs

File: packages/contact-center/store/package.json [hunk 7]
```diff
@@ -37,6 +44,7 @@
     "webpack-merge": "6.0.1"
   },
   "jest": {
-    "testEnvironment": "jsdom"
+    "testEnvironment": "jsdom",
+    "passWithNoTests": true
```
Comment (Shreyas281299) [inline packages/contact-center/store/package.json:48]: We can remove this when we have tests

File: .circleci/config.yml [hunk 8]
```diff
@@ -98,19 +120,32 @@ workflows:
...
               only: master
+              
+
+      - release_widgets:
+          requires:
+            - build
+            - unit_tests
+          filters:
+            branches:
+              only: feat/cc-widgets
```
Comment (Shreyas281299) [inline .circleci/config.yml:151]: This make sure we only release ccwidgets when we push onto feat/cc-widgets branch

File: .releaserc [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+      {
+        "changelogFile": "packages/contact-center/CHANGELOG.md"
+      }
+    ],
+    [
+      "@semantic-release/exec",
+      {
+        "prepareCmd": "node ./tooling/src/publish.js ${nextRelease.channel} ${nextRelease.version} "
+      }
+    ],
```
Comment (Shreyas281299) [inline .releaserc:23]: Leaving this commit to highlight that this part needs to be added once we have added 2 releases onto github.

File: tooling/tests/publish.js [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+    process.argv = ['node', 'script.js', 'main', '1.3.3-test.1'];
+    versionAndPublish();
+
+    const expectedPackageJsonContent = JSON.stringify({name: 'test-package', version: '1.0.0'}, null, 2);
+
+    expect(mockFs.writeFileSync).toHaveBeenCalledWith(packageJsonPath, expectedPackageJsonContent, 'utf-8');
+    expect(console.log).toHaveBeenCalledWith("'stableVersion' key removed successfully.");
+  });
+
+  test('error occured while writting "stableVersion" key to', () => {
```
Comment (mkesavan13) [inline tooling/tests/publish.js:58]: Check typos and write full statement

**Review by mkesavan13 — CHANGES_REQUESTED** [review]

File: tooling/tests/publish.js (hunk 10 above)
Comment (Shreyas281299) [reply tooling/tests/publish.js]: Updated the statement

File: tooling/src/publish.js [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+      .readdirSync(contactCenterPath, {withFileTypes: true})
+      .filter((dirent) => dirent.isDirectory())
+      .map((dirent) => dirent.name);
+    // Separate dependency workspaces and other workspaces
+    const dependencyWorkspaces = ccFolder.filter((fileName) => dependencies.includes(`@webex/cc-${fileName}`));
+
+    const otherWorkspaces = ccFolder.filter((fileName) => !dependencies.includes(`@webex/cc-${fileName}`));
+
+    const publishWorkspace = (fileName) => {
+      const workspaceName = `@webex/cc-${fileName}`;
```
[the remark this answers is not in the document]
Comment (Shreyas281299) [reply tooling/src/publish.js]: Updated the logic to get name from package.json

File: packages/contact-center/user-state/package.json [hunk 12]
```diff
@@ -38,6 +45,7 @@
     "webpack-merge": "6.0.1"
   },
   "jest": {
-    "testEnvironment": "jsdom"
+    "testEnvironment": "jsdom",
+    "passWithNoTests": true
```
[the remark this answers is not in the document]
Comment (Shreyas281299) [reply packages/contact-center/user-state/package.json:49]: Added

File: packages/contact-center/store/package.json (hunk 7 above)
Comment (Shreyas281299) [reply packages/contact-center/store/package.json:48]: Updated

**Review by mkesavan13 — APPROVED** [review]

**Comment by Shreyas281299** [comment]
Screenshot 2024-11-21 at 11 44 37 AM

Changed the pipeline to add release on github with the formate `v${version}`.

File: package.json [hunk 13]
```diff
@@ -11,20 +11,38 @@
...
     "webpack": "^5.96.1",
     "webpack-cli": "^5.1.4",
     "webpack-dev-server": "^5.1.0"
   },
   "scripts": {
+    "test:unit": "yarn run test:tooling && yarn run test:cc-widgets",
+    "test:tooling": "jest --coverage",
+    "test:cc-widgets": "yarn run test:store && yarn run test:widgets",
+    "test:store": "yarn workspace @webex/cc-store test:unit",
+    "test:widgets": "yarn workspace @webex/cc-station-login test:unit && yarn workspace @webex/cc-user-state test:unit",
```
Comment (sreenara) [inline package.json:33]: This is a suggestion for improvement. We could use `yarn workspaces foreach` when we'd like to run the same command across different packages

**Pull request merged**